File: Python/Heap/Binary_MaxHeap.py
#---------------------------------Binary Heap Time Complexity------------------------------------
#  Insertion - O(logn)
#  Removal   - O(logn)
#  Searching - O(n)    

import math

class MaxBinaryHeap:
    def __init__(self):
        self.values = []
    
    def insert(self, val):
        self.values.append(val)
        self.BubbleUp()

    def BubbleUp(self):
        index = len(self.values) - 1
        element = self.values[index]
        while index > 0:
            parentIndex = math.floor(index/2);
            parentElement = self.values[parentIndex]
            if element <= parentElement:
                break
            self.values[parentIndex] = element
            self.values[index] = parentElement
            index = parentIndex
    
    def extractMax(self):
        max = self.values[0]
        end = self.values.pop()
        if len(self.values) > 1:
            self.values[0] = end
            self.bubbleDown()
        return max


    def bubbleDown(self):
        index = 0
        length = len(self.values)
        element = self.values[0];
        while True:
            leftIndex = 2*index + 1
            rightIndex = 2*index + 2
            leftChild, rightChild = None, None
            swap = None
            if leftIndex < length:
                leftChild = self.values[leftIndex]
                if leftChild > element:
                    swap = leftIndex
            
            if rightIndex < length:
                rightChild = self.values[rightIndex]
                if (swap == None and rightChild > element) or (swap != None and rightChild > leftChild):
                    swap = rightIndex
            
            if swap == None:
                break

            self.values[index] = self.values[swap]
            self.values[swap] = element
            index = swap 

heap = MaxBinaryHeap()
print(heap.insert(41))
print(heap.insert(39))
print(heap.insert(33))
print(heap.insert(18))
print(heap.insert(27))
print(heap.insert(12))
print(heap.values)
print(heap.extractMax())
print(heap.values)
print(heap.extractMax())
print(heap.values)

File: Js/Heap/Binary_MaxHeap.js
class MaxBinaryHeap{
    constructor(){
        this.values = [];
    };

    insert(val){
        this.values.push(val);
        this.bubbleUp();
    };

    bubbleUp(){
        let index = this.values.length - 1;
        const element = this.values[index];
        while(index > 0){
            let parentIndex = Math.floor((index-1)/2);
            let parent = this.values[parentIndex];
            if(element <= parent) break;
            this.values[parentIndex] = element;
            this.values[index] = parent
            index = parentIndex
        }        
    }

    extractMax(){
        const max = this.values[0];
        const end = this.values.pop();
        if(this.values.length > 1){
            this.values[0] = end;
            this.bubbleDown();
        }
        return max;
    }

    bubbleDown(){
        let index = 0;
        const length = this.values.length;
        const element = this.values[0];
        while(true){
            let leftIndex = 2 * index + 1;
            let rightIndex = 2 * index + 2;
            let leftChild,rightChild;
            var swap = null;
            if(leftIndex < length){
                leftChild = this.values[leftIndex];
                if(leftChild > element){
                    swap = leftIndex;
                }
            }
            if(rightIndex < length){
                rightChild = this.values[rightIndex];
                if(
                    (swap === null && rightChild > element) ||
                    (swap !== null && rightChild > leftChild)
                ){
                    swap = rightIndex;
                }
            }
            if(swap === null)   break;

            this.values[index] = this.values[swap];
            this.values[swap] = element;
            index = swap;
        }
    }
};

let heap = new MaxBinaryHeap();
heap.insert(41)
heap.insert(39)
heap.insert(33)
heap.insert(18)
heap.insert(27)
heap.insert(12)
heap.insert(55)
console.log(heap.values);
console.log(heap.extractMax());
console.log(heap.values);